File: 0137-single-number-ii/0137-single-number-ii.java
class Solution {
    public int singleNumber(int[] nums) {
        int ans = 0;
        for (int i = 0; i < 32; i++) {
            int ones = 0;
            for (int num : nums) {
                if ((num & (1 << i)) != 0) {
                    ones++;
                }
            }
            if (ones % 3 != 0) {
                ans |= (1 << i);
            }
        }
        return ans;
    }
}


// class Solution {
//     public int singleNumber(int[] nums) {
//         int ones = 0, twos = 0;
//         for (int num : nums) {
//             ones = (ones ^ num) & ~twos;
//             twos = (twos ^ num) & ~ones;
//         }
//         return ones;
//     }
// }

File: LeetCode/0137-single-number-ii/0137-single-number-ii.py
class Solution:
    def singleNumber(self, nums: List[int]) -> int:
        ans = 0
        for i in range(32):
            ones = 0
            for num in nums:
                if num & 1 << i:
                    ones += 1
            if ones % 3 != 0:
                ans |= 1 << i
        return ans

# class Solution:
#     def singleNumber(self, nums: List[int]) -> int:
#         ones, twos, threes = 0, 0, 0
#         for num in nums:
#             twos |= ones & num
#             ones ^= num
#             threes = ones & twos
#             ones &= ~threes
#             twos &= ~threes
#         return ones